### ISSUE TITLE
AMP.pushState() allow change URL as history.pushState(stateObj, title, URL);

### ISSUE THREAD

**Issue by eduardogoncalves** [ampproject/amphtml#23560]
At `history.pushState` we can pass `URL` `history.pushState(stateObj, title, URL);` to change URL in browser without navigating away from page.

Could AMP implement it?

Ex.: In a SPA, an user select Tweets tab, and url change to `/tweets`, then select Agenda and url change to `/agenda`.

![image](https://user-images.githubusercontent.com/4308648/61996787-7ce44c00-b066-11e9-87a9-d0b6c36735d4.png)

**Comment by nainar** [comment]
@cvializ worthy to keep in mind as part of stateful AMP pages.

**Comment by cvializ** [comment]
I think some aspects of this FR are already implemented in amp-bind with the `AMP.pushState` and `AMP.popState` actions. https://amp.dev/documentation/components/amp-bind/?referrer=ampproject.org#modifying-history-with-amp.pushstate()

**Comment by choumx** [comment]
That's correct. Also we'd need viewer coordination to update title and hash.

**Comment by chasefinch** [comment]
+1 At BringFido, title/url updating on state change (when not being shown from an amp cache) is the last mission-critical thing we can't pull off in AMP.

**Comment by choumx** [comment]
The original API proposal: https://github.com/ampproject/amphtml/issues/9128#issuecomment-320332045

**Comment by bracketdash** [comment]
So has anyone had luck with getting URLs to change without this issue being addressed, or are we all pretty much waiting for this?

**Comment by stephengardner** [comment]
From an AMP-first perspective, many websites have a search/filter functionality which updates the URL so that it can be easily shared with filters pre-applied.

Similarly, all Shopify websites update the URL with ?variant={{variant.id}} so that when sharing a page with friends, the recipient auto-selects the proper variant.

For websites that do this, it does seem especially useful.

Is this intended on being implemented any time soon?

**Comment by choumx** [comment]
Definitely possible, just hasn't been prioritized yet. 

> Similarly, all Shopify websites update the URL with ?variant={{variant.id}}

I think this can be done today with `amp-list` and a `QUERY_PARAM` variable substitution in the `src` attribute, though it would fire a network request.

**Comment by dennisP2ER** [comment]
This feature would be really awesome.

**Comment by AaronMLB** [comment]
This would be extremely useful if it could be finished and rolled out. Between some of the comments and the existing ```amp-bind``` ```AMP.pushState()``` functionality it appears many of the fundamentals have been tackled.

Any chance of being re-prioritised, please? It feels like one of those small steps that actually would make AMP that larger step more usable and comprehensive in the real world.

**Comment by chasefinch** [comment]
Bump

**Comment by digitalheir** [comment]
https://twitter.com/marcan42/status/1581244983528820741?cxt=HHwWisDUoen72vErAAAA
